File: frontend/iblog-site-frontend/src/api/ws/WSChatMessage.ts
export class WSChatMessage {
    public id: string;
    public groupId: string;
    public type: WSChatMessageType;
    public chatContentMessage: WSChatContentMessage;
    public userJoinMessage: WSUserJoinMessage;
    public createdTime: Date;


    constructor(id: string, groupId: string, type: WSChatMessageType, chatContentMessage: WSChatContentMessage, userJoinMessage: WSUserJoinMessage, createdTime: Date) {
        this.id = id;
        this.groupId = groupId;
        this.type = type;
        this.chatContentMessage = chatContentMessage;
        this.userJoinMessage = userJoinMessage;
        this.createdTime = createdTime;
    }
}

export class ChatMember {
    public userId: number;
    public name: string;


    constructor(userId: number, name: string) {
        this.userId = userId;
        this.name = name;
    }
}

export enum WSChatMessageType {
    USER_JOIN = "USER_JOIN",
    CHAT = "CHAT"
}

export class WSUserJoinMessage {
    public chatMember: ChatMember;

    constructor(chatMember: ChatMember) {
        this.chatMember = chatMember;
    }
}

export class WSChatContentMessage {
    public from: ChatMember;
    public to: ChatMember | null;
    public content: string;

    constructor(from: ChatMember, to: ChatMember | null, content: string) {
        this.from = from;
        this.to = to;
        this.content = content;
    }
}

File: frontend/iblog-site/src/main/java/app/site/cache/WSChatMessage.java
package app.site.cache;

import org.springframework.data.annotation.Id;
import org.springframework.data.redis.core.RedisHash;
import org.springframework.data.redis.core.index.Indexed;

import java.io.Serializable;
import java.time.ZonedDateTime;
import java.util.Arrays;

/**
 * @author steve
 */
@RedisHash(value = "messages")
public class WSChatMessage implements Serializable {
    @Id
    public String id;

    @Indexed
    public String groupId;

    @Indexed
    public WSChatMessageType type;

    public WSChatContentMessage chatContentMessage;

    public WSUserJoinMessage userJoinMessage;

    public ZonedDateTime createdTime;

    public static class ChatMember {
        public Long userId;

        public String name;
    }

    @Override
    public String toString() {
        return "WSChatMessage{" +
            "id='" + id + '\'' +
            ", groupId='" + groupId + '\'' +
            ", type=" + type +
            ", chatContentMessage=" + chatContentMessage +
            ", userJoinMessage=" + userJoinMessage +
            ", createdTime=" + createdTime +
            '}';
    }
}
